Improve: Match detached window menu to main and focus behavior

mpconley opened the pull request:
#### Brief overview of PR changes/additions

Transforms detached windows into fully functional independent workspaces with complete menu systems and proper focus management. When dialogs are closed, focus now correctly returns to the originating profile tab instead of jumping to the main window. Also fixes profile management issues where reconnecting would restore profiles to unexpected window locations.

#### Motivation for adding to Mudlet

Detached windows were previously limited - users had to constantly switch back to the main window for most operations and experienced confusing focus behavior when closing dialogs. Profile management was also broken, with closed profiles sometimes reappearing in the wrong window after reconnection. This made multi-profile workflows inefficient and frustrating.

#### Other info (issues closed, discussion etc)

- **Complete menu system**: Added Games, Toolbox, Options, Window, Help, and About menus to detached windows with full feature parity
- **Universal focus restoration**: All dialogs (script editors, notepad, package manager, module manager, preferences, package exporter) now properly restore focus to the correct profile tab
- **Fixed profile cleanup**: Profiles no longer appear in wrong windows after disconnection and reconnection
- **Code consolidation**: Eliminated significant duplicate code while improving maintainability
- **UI consistency**: Added missing options like "Show Connection Indicators" to main window for parity
- Closes issue #8192 (detached window toolbox accessibility) and #8109 (`Window` menu consistency)

mpconley committed:
Fix detached window profile cleanup and add missing context menu option

- Fix profile cleanup: Remove detached window mapping when profiles are closed
  to prevent recreated profiles from appearing in wrong window
- Add 'Show Connection Indicators on Tabs' option to main window context menu
  for parity with detached windows
- Ensure new profiles always appear in main window after connection dialog
- Clean up detached window tracking in mudlet::closeHost()

Fixes issues where:
1. Closing a profile in detached window then reconnecting would put the
   profile back in the detached window instead of main window
2. Main window tabs were missing the connection indicator toggle option
   that was available in detached windows

mpconley committed:
Implement centralized focus restoration for script editor dialogs in detached windows

This commit implements a robust, centralized focus restoration system for all script editor dialogs opened from detached windows. The solution addresses the issue where closing script editors would not properly restore focus to the originating profile tab in detached windows.

Key Changes:
- Added custom 'editorClosing' signal to dlgTriggerEditor for reliable close detection
- Implemented centralized setupEditorFocusRestoration() function in mudlet.cpp
- Updated all script editor dialog methods in TDetachedWindow to use the new pattern:
  * slot_showTriggerDialog() (already updated)
  * slot_showScriptDialog() (already updated)
  * slot_showAliasDialog()
  * slot_showTimerDialog()
  * slot_showActionDialog()
  * slot_showKeyDialog()
  * slot_showVariableDialog()
  * slot_showEditorDialog()

Technical Improvements:
- Replaced unreliable 'destroyed' signal with custom 'editorClosing' signal
- Context-aware focus restoration that works for both main and detached windows
- Direct editor creation to avoid interference from main window focus logic
- Consistent pattern across all script editor dialogs
- Removed complex manual focus restoration code and QTimer workarounds

The implementation ensures that when any script editor is closed, focus is properly restored to the correct profile tab in the detached window where it was opened from, providing a seamless user experience across all scripting tools.

mpconley committed:
Consolidate duplicate script editor dialog code

This commit eliminates significant code duplication by consolidating all script editor dialog methods into a single helper function. All script editor slot methods now use the same centralized pattern for consistency and maintainability.

Key Changes:
- Added showScriptEditorDialog() helper method that takes a function parameter
- Simplified all slot_show*Dialog() methods to 3-4 lines each:
  * slot_showTriggerDialog()
  * slot_showScriptDialog()
  * slot_showAliasDialog()
  * slot_showTimerDialog()
  * slot_showActionDialog()
  * slot_showKeyDialog()
  * slot_showVariableDialog()
  * slot_showEditorDialog()

Benefits:
- Reduced code duplication from ~40 lines per method to ~4 lines per method
- Single point of maintenance for script editor dialog logic
- Consistent behavior across all script editor types
- Same reliable focus restoration for all dialogs
- Easier to add new script editor dialog types in the future

The helper method encapsulates all the common logic (profile capture, editor creation/retrieval, focus restoration setup, showing/activation) while allowing each dialog type to specify its specific display method via a lambda function.

mpconley committed:
Implement focus restoration for Notepad dialog

- Add notepadClosing signal to dlgNotepad that emits on closeEvent
- Implement setupNotepadFocusRestoration in mudlet.cpp to handle focus restoration
- Support both detached windows and main window focus restoration
- Automatically set up focus restoration when notepad is created for a profile
- Use same pattern as script editor focus restoration but adapted for persistent dialogs

This ensures that when a Notepad is closed, focus returns to the correct
profile tab whether in the main window or a detached window.

mpconley committed:
Consolidate focus restoration logic for reusability

- Extract common focus restoration logic into restoreProfileFocus() function
- Simplify setupEditorFocusRestoration to use common logic when no target window specified
- Simplify setupNotepadFocusRestoration to use common logic
- Maintain backward compatibility for setupEditorFocusRestoration's target window parameter
- Create foundation for easily adding focus restoration to other dialogs

This consolidation reduces code duplication and provides a clear pattern for
extending focus restoration to additional dialog types in the future.

mpconley committed:
xtend focus restoration to Package Manager, Module Manager, Package Exporter, and Preferences dialogs

- Added custom signals (packageManagerClosing, moduleManagerClosing, packageExporterClosing, preferencesClosing) to all dialog classes
- Added closeEvent overrides to emit signals when dialogs are closed
- Added QCloseEvent includes to header files
- Created setup*FocusRestoration functions for each dialog type
- Connected setup functions to dialog creation in mudlet.cpp slot methods
- Extended the existing focus restoration pattern to work consistently across all profile-related dialogs
- All dialogs now restore focus to the correct profile tab when closed, whether from main or detached windows

mpconley committed:
Merge branch 'development' into orphan_detached_toolbox

mpconley commented on src/mudlet.h:
```diff
@@ -112,7 +112,13 @@ class QTimer;
+class dlgModuleManager;
+class dlgPackageExporter;
 class dlgProfilePreferences;
+class dlgProfilePreferences;
```
Duplicate

mpconley commented on src/dlgProfilePreferences.cpp:
```diff
@@ -4600,3 +4600,11 @@ void dlgProfilePreferences::slot_changeShowTabConnectionIndicators(bool state)
+void dlgProfilePreferences::closeEvent(QCloseEvent* event)
+{
+    if (mpHost) {
+        emit preferencesClosing(mpHost->getName());
```
Check what happens if no profile yet / still exists.

mpconley committed:
Fix Ubuntu build: Wrap updater methods in INCLUDE_UPDATER ifdef

The update check and changelog menu items in detached windows were calling
methods that are only available when INCLUDE_UPDATER is defined. This caused
build failures on platforms where the updater is not included.

Wraps the update and changelog action creation and connection in
INCLUDE_UPDATER ifdefs to match the pattern used in the main window.

mpconley committed:
Remove duplicate dlgProfilePreferences forward declaration

Removes duplicate forward declaration that was accidentally added during
the focus restoration implementation. Keeps the codebase clean and
prevents potential compilation warnings.

ZookaOnGit committed:
Merge branch 'development' into orphan_detached_toolbox

mpconley committed:
Merge branch 'development' into orphan_detached_toolbox

mpconley commented:
@ZookaOnGit for this one, given there are so many menu items, it needs someone to try them out and confirm it works like the main window menus. Outside of that what this PR does is the effort needed to ensure when the dialogs get opened from the main/detached window that it brings focus back to the profile it was opened when closed.

mpconley committed:
Update detached window shortcuts to match main window platform logic

- Apply platform-conditional shortcut logic to all detached window menu actions
- Use CTRL+key on macOS, ALT+key on other platforms (consistent with main window)
- Ensures cross-platform menu shortcut consistency between main and detached windows
- Removes hardcoded shortcuts that didn't match platform conventions

sammerpetria commented:
Hi, tested with 3 monitors.. lot of profiles, multiview...  and works very good!  this on windows 11.

Maybe a thing, if in second monitor for example, the editor opens in the main profile monitor.

First i think thats better for front of view focus. but when u open two at the time requieres to read good the name to not edit the wrong one. Maybe load in the same monitor has the profile is would be better. So if you stay in monitor 2 with one game and open editor is all in 2nd monitor then monitor 1 another... etc gives consitency. If not..just 1 second to mouse move is not a problem.

Like detach a lot great job

SlySven commented:
@sammerpetria Unless things have changed we do not store the editor position on a "per-profile" basis but only the position of the editor for the last profile that edited (IIRC). The values are contained in the application-wide settings (now in the `Mudlet.ini` file for all OS) under the `["General"]` heading/section as `script_editor_pos` and `script_editor_size`.

TBH It has been something (but which was very low down) on my TODO list to instead move the per-profile values (with names something like `editorPos` and `editorSize`) to the `profile.ini` file in the root of each profile part of the file-system.

However that is something that should be deferred to a separate PR.

mpconley committed:
Fix: Ensure dialogs open on correct monitor for multi-monitor setups

- Add utility functions for smart dialog positioning in utils.h
- Scripts dialog now uses screen-aware positioning instead of saved positions
- Dialog positioning considers active profile location (main window vs detached)
- Force repositioning on dialog show to handle profile reattachment scenarios
- Singleton dialogs (Scripts, Preferences, etc.) now consistently follow active profile
- Fixes issue where dialogs opened on wrong monitor after profile reattachment

The Scripts dialog and other singleton dialogs now properly track the active
profile's location and open on the same monitor, regardless of saved position
data or profile detachment/reattachment operations.

mpconley commented:
Great feedback @sammerpetria and @SlySven, this helped me understand the problem and part of the solution. If you'd like to test, the editors, notepads, etc. should follow the profile to its monitor.

1. Open profile 1 in monitor 1, open its script editor, close its script editor.
2. Open profile 2 in monitor 1, open its script editor, close its script editor.
3. Detach profile 2 in monitor 1, open its script editor, close its script editor.
4. Drag the detached window to monitor 2, open its script editor, close its script editor.
5. Reattach profile 2 to the main window.
6. Open profile 2 in monitor 1, open its script editor, close its script editor.
7. Detach profile 2 in monitor 1, open its script editor, close its script editor.

This video captured monitor 1, I think.

https://github.com/user-attachments/assets/975f3ea7-f09b-4336-82b3-590afa886404

mpconley committed:
Merge branch 'development' into orphan_detached_toolbox

mpconley committed:
Fix: CodeFactor formatting issue in utils.h

Move closing parenthesis to previous line in QPoint constructor to match
coding style guidelines (whitespace/parens).

mpconley committed:
Improve: Use consistent mudletInstance for focus restoration calls

Replace mudlet::self() calls with mudletInstance in TDetachedWindow dialog methods:
- slot_showPreferencesDialog(): setupPreferencesFocusRestoration()
- slot_showNotesDialog(): setupNotepadFocusRestoration()
- slot_showPackageManagerDialog(): setupPackageManagerFocusRestoration()
- slot_showModuleManagerDialog(): setupModuleManagerFocusRestoration()

This improves code consistency and safety by using the already-retrieved
mudletInstance variable with null checking instead of making additional
unchecked mudlet::self() calls.

sammerpetria commented:
@mpconley tested up to 3 monitors at time, 3 profiles, works as supposed to work! 
this on windows 11.
nice job

mpconley committed:
Merge branch 'development' into orphan_detached_toolbox

vadi2 reviewed (approved):
Thanks for this!

The method new names in utils are a bit lengthy, but I couldn't see anything shorter either.

🚀

mpconley commented:
Ready to merge? Thanks.
